### PR TITLE
Fold E0612, E0613 into E0609

### LINKED ISSUE #42945: E0613 needs to be merged into E0609 now
With RFC 1506 merged and stabilized, handling `E0613` separate from `E0609` is not the correct behaviour any more.

For context, you get `E0613` when doing something like `let _ = "hi".0;`:

```
error[E0613]: attempted to access tuple index `0` on type `&'static str`, but the type was not a tuple or tuple struct
```

And you get `E0609` when doing something like `let _ = "hi".foo;`:

```
error[E0609]: no field `foo` on type `&'static str`
```

[RFC 1506 "clarified ADT kinds"](https://github.com/rust-lang/rfcs/pull/1506) ([Rendered](https://github.com/rust-lang/rfcs/blob/master/text/1506-adt-kinds.md), [tracking issue](https://github.com/rust-lang/rust/issues/35626)) allowed all structs to contain fields with integral names like `0` or `1`, etc. This makes the error message for `E0613` outdated, as the type may be a non tuple struct, and calling `0` a "tuple index" sounds wrong as well.

So I propose `E0613` to be merged into `E0609`, as `E0609` is now better suiting, and also because its "in the spirit" of the RFC.

The proposed new output for `let _ = "hi".0;` is:
 
```
error[E0609]: no field `0` on type `&'static str`
```

### PR DESCRIPTION
**Pull request by strangeglyph**
As discussed in #42945, with PR 1506 tuple indices are no longer considered a separate case from normal field. This PR folds E06012 ("tuple index out of bounds") and E0613 ("type is not a tuple") into E0609 ("type does not have field with that name")

Resolves #42945

### CODE REVIEW

**Comment by Boreeas** [comment]
My apologies, I forgot to run the full test suite and apparently missed something.

**Comment by Boreeas** [comment]
All tests pass now on my end

**Comment by Mark-Simulacrum** [comment]
Thanks for the pull request! We'll get @pnkfelix or someone else from the compiler team to review this soon.

cc @GuillaumeGomez -- perhaps you'd like to take a look since I believe you created some of these in the first place?

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3057,20 +3057,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             return field_ty;
         }
 
-        if tuple_like {
-            type_error_struct!(self.tcx().sess, expr.span, expr_t, E0612,
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: I feel like this error is a bit different than the others. Not sure if this is a good idea to merge it as well...

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3057,20 +3057,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                               idx.node, expr_t).emit();
-        }
-
+        self.no_such_field_err(expr.span, idx.node, expr_t).emit();
         self.tcx().types.err
     }
 
+    fn no_such_field_err<T: Display>(&self, span: Span, field: T, expr_t: &ty::TyS)
+        -> DiagnosticBuilder {
+
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: Extra empty line.

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -4419,83 +4419,8 @@ println!("{}", y.get()); // So we can get the value through the function.
...
-}
-
-let s = SomeStruct {
-    x: 0,
-    y: -1,
-};
-println!("x: {} y: {}", s.x, s.y);
-```
-"##,
+// E0612: Removed (merged with E0609)
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Please put them into the error code list at the bottom.

File: src/test/compile-fail/issue-19244-1.rs [hunk 4]
```diff
@@ -12,5 +12,5 @@ const TUP: (usize,) = (42,);
 
 fn main() {
     let a: [isize; TUP.1];
-    //~^ ERROR attempted out-of-bounds tuple index
+    //~^ ERROR no field `1` on type `(usize,)`
```
Comment (GuillaumeGomez) [inline src/test/compile-fail/issue-19244-1.rs]: This error is clearly less good from my point of view. Before it was very clear but now, seems more confusing.

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
I really think we should keep E0612 on its own and not merge it with others.

**Comment by Boreeas** [comment]
@GuillaumeGomez Alright, reverted the changes to E0612

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks! Please [squash](https://blog.guillaume-gomez.fr/Git-tips/1/2) your commits and then it's good for me.

**Comment by alexcrichton** [comment]
ping @GuillaumeGomez, looks like this has been updated and is ready for a re-review?

**Comment by GuillaumeGomez** [comment]
The review was done, I was waiting for the squash. Since it's done, we can r+. Thanks @Boreeas!

@bors: r+

**Comment by Boreeas** [comment]
No problem! My apologies for not sending a notice that I squashed

**Comment by Boreeas** [comment]
Rebased the PR

**Comment by GuillaumeGomez** [comment]
Then here we go again!

@bors: r+

**Pull request merged**